### PR TITLE
Rianfowler/sc 28650/kotsadm validation of config options entered

### PR DESCRIPTION
**Pull request by rianfowler**
#### What this PR does / why we need it:
- adds regex validation and error rendering to text, textarea, and password fields. 
- https://www.loom.com/share/fcea5606ef1f40cca2db6670506b1c09
Screenshot 2023-04-14 at 10 57 09

#### Which issue(s) this PR fixes:

Fixes # https://app.shortcut.com/replicated/story/28650/kotsadm-validation-of-config-options-entered-by-user

#### Does this PR introduce a user-facing change?

```release-note
admin console: adds support for showing custom validation errors on config
```

#### Does this PR require documentation?

https://app.shortcut.com/replicated/story/73619/orange-team-regex-config

### CODE REVIEW

File: web/src/components/config_render/ConfigGroup.jsx
Comment (rianfowler) [inline web/src/components/config_render/ConfigGroup.jsx:1]: We are getting a regex and an error message per field from the backend. The `showValidationError` is a state the FE adds in the `AppConfig` when the config response is received.

File: web/src/components/config_render/ConfigInput.jsx [hunk 1]
```diff
@@ -202,13 +202,19 @@ export default class ConfigInput extends React.Component {
               this.props.readonly ? "readonly" : ""
             } tw-gap-0`}
             isFirstChange={this.state.isFirstChange}
+            showError={this.props.showValidationError}
```
Comment (rianfowler) [inline web/src/components/config_render/ConfigInput.jsx:205]: Updated in the `<InputField />` to handle this.

File: web/src/components/config_render/ConfigInput.jsx [hunk 2]
```diff
@@ -202,13 +202,19 @@ export default class ConfigInput extends React.Component {
...
         {this.props.inputType !== "password" && this.props.default ? (
           <div className="default-value-section u-marginTop--8">
             Default value: <span className="value"> {this.props.default} </span>
           </div>
         ) : null}
+        {this.props.showValidationError && (
+          <div className="config-input-error-message tw-mt-1 tw-text-xs">
+            {this.props.validationErrorMessage}
+          </div>
+        )}
```
Comment (rianfowler) [inline web/src/components/config_render/ConfigInput.jsx:217]: shows the error message under the input field

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 3]
```diff
@@ -427,6 +435,26 @@ class AppConfig extends Component<Props, State> {
...
+    }
+
+    const validationRegex = new RegExp(validation.regex);
+
+    if (validationRegex.test(value)) {
+      return { ...validation, hasError: false };
+    } else {
+      return { ...validation, hasError: true };
+    }
+  };
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:457]: This is kind of clunky but it takes a Validation and a value and returns a Validation. This function primarily modifies `hasErrors`

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 4]
```diff
@@ -480,6 +508,20 @@ class AppConfig extends Component<Props, State> {
...
               }
             }
+            // test data for regex
+            if (["text", "password", "textarea"].includes(newItem.type)) {
+              // TODO: remove- this is used for generating test data locally
+              // newItem.validation = {
+              //   regex: "^.{4,20}$",
+              //   message: "should be between 4 and 20 characters",
+              //   hasError: false,
+              // };
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:522]: No BE support for this so i'm leaving my test data generation stuff in for now. This adds a length constraint to every text password and textarea field

File: web/src/components/config_render/ConfigRender.jsx
Comment (rianfowler) [inline web/src/components/config_render/ConfigRender.jsx:1]: These changes are mostly unrelated but I noticed the debouncing probably doesn't work as intended so I updated it here.

this.props.getData actually updates the entire config state and returns nothing. I didn't change that but I did move that call into the debounced function.

**Review by miaawong — DISMISSED** [review]

File: web/src/components/config_render/ConfigFileInput.jsx [hunk 5]
```diff
@@ -132,6 +132,11 @@ export default class ConfigFileInput extends React.Component {
                 }
               />
             </span>
+            {this.props.showValidationError && (
+              <div className="config-input-error-message tw-mt-1 tw-text-xs">
+                {this.props.validationErrorMessage}
+              </div>
+            )}
```
Comment (rianfowler) [inline web/src/components/config_render/ConfigFileInput.jsx:139]: shows validation error on input

File: web/src/components/config_render/ConfigGroup.jsx [hunk 6]
```diff
@@ -58,6 +58,8 @@ const ConfigGroup = (props) => {
               {...item}
               readonly={isReadOnly}
               index={i + 1}
+              validationErrorMessage={item?.validationError}
+              showValidationError={item?.validationError}
```
Comment (rianfowler) [inline web/src/components/config_render/ConfigGroup.jsx:62]: keeping these as separate props for now

File: web/src/components/config_render/ConfigRender.jsx
Comment (rianfowler) [inline web/src/components/config_render/ConfigRender.jsx:1]: Unrelated refactor. `triggerChange` is debounced but the actual work is happening in `getData` so moved that into the debounced behavior rather than having `getData` passed in as an argument (where it is not debounced).

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 7]
```diff
@@ -38,12 +38,23 @@ type Props = {
...
+interface ConfigGroupItemValidationErrors {
+  item_errors: {
+    name: string;
+    validation_errors: {
+      message: string;
+    }[];
+  }[];
+  name: string;
+}
+
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:57]: new root level property on `liveconfig` response

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 8]
```diff
@@ -38,12 +38,23 @@ type Props = {
 // This was typed from the implementation of the component so it might be wrong
 type ConfigGroup = {
   hidden: boolean;
+  hasError: boolean;
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:41]: This is used to auto expand the nav groups

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 9]
```diff
@@ -465,7 +479,43 @@ class AppConfig extends Component<Props, State> {
 
         const data = await response.json();
         const oldGroups = this.state.configGroups;
-        const newGroups = data.configGroups;
+        const validationErrors: ConfigGroupItemValidationErrors[] =
+          data.validationErrors;
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:483]: The new validationError property is returned at the root level of the response

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 10]
```diff
@@ -465,7 +479,43 @@ class AppConfig extends Component<Props, State> {
...
+
+        // track errors at the form level
+        this.setState({ showValidationError: false });
+        let hasValidationError = false;
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:487]: We use this to set the error by the save config button and disable it. Since `setState` is async we need the local var `hasValidationError` to prevent multiple `setStates` from being called

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 11]
```diff
@@ -465,7 +479,43 @@ class AppConfig extends Component<Props, State> {
...
+            newGroup.items = newGroup.items.map((item: ConfigGroupItem) => {
+              const itemValidationError =
+                configGroupValidationErrors.item_errors.find(
+                  (validationError) => validationError.name === item.name
+                );
+
+              if (itemValidationError && item.value) {
+                item.validationError =
+                  itemValidationError.validation_errors[0].message;
+                newGroup.hasError = true;
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:506]: This is the actual meat of the change. If there is an error then we add it as `validationError` to the item and `hasError` to the group. 

The rest of the logic is just querying the new `validationErrors` returned from the API and matching the `group.name` and `item.name` that links the `validationError` item to the group item.

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 12]
```diff
@@ -647,7 +698,8 @@ class AppConfig extends Component<Props, State> {
                 <div
                   key={`${i}-${group.name}-${group.title}`}
                   className={`side-nav-group ${
-                    this.state.activeGroups.includes(group.name)
+                    this.state.activeGroups.includes(group.name) ||
+                    group.hasError
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:702]: auto expand errors in the side nav

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 13]
```diff
@@ -682,11 +734,13 @@ class AppConfig extends Component<Props, State> {
                           }
                           return (
                             <a
-                              className={`u-fontSize--normal u-lineHeight--normal ${
-                                hash === `${item.name}-group`
-                                  ? "active-item"
-                                  : ""
-                              }`}
+                              className={`u-fontSize--normal u-lineHeight--normal
+                                ${item.validationError ? "has-error" : ""}
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:738]: make the errors in the side nav show up as red

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 14]
```diff
@@ -748,14 +802,17 @@ class AppConfig extends Component<Props, State> {
...
-                                {configError}
-                              </span>
-                            )}
+                            {configError ||
+                              (this.state.showValidationError && (
+                                <span className="u-textColor--error tw-mb-2 tw-text-xs">
+                                  {configError ||
+                                    "Error detected. Please use config nav to the left to locate and fix issues."}
+                                </span>
+                              ))}
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:811]: show error above button. 

This error already exists but I tweaked it to show the validation error if there's no configError. I also reduced the size so that it shows up like the rest of the errors.

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 15]
```diff
@@ -465,7 +479,43 @@ class AppConfig extends Component<Props, State> {
...
+            newGroup.items = newGroup?.items?.map((item: ConfigGroupItem) => {
+              const itemValidationError =
+                configGroupValidationErrors?.item_errors?.find(
+                  (validationError) => validationError.name === item.name
+                );
+
+              if (itemValidationError && item.value) {
+                item.validationError =
+                  itemValidationError?.validation_errors?.[0]?.message;
+                newGroup.hasError = true;
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:506]: This is the actual meat of the change. If there is an error then we add it as validationError to the item and hasError to the group.

The rest of the logic is just querying the new validationErrors returned from the API and matching the group.name and item.name that links the validationError item to the group item.

File: web/src/features/AppConfig/components/AppConfig.tsx [hunk 16]
```diff
@@ -427,6 +481,76 @@ class AppConfig extends Component<Props, State> {
     return foundItem;
   };
 
+  // this runs on config update and when save is clicked but before the request is submitted
+  // on update it uses the errors from the liveconfig endpoint
+  // on save it's mostly used to find required field errors
+  mergeConfigGroupsAndValidationErrors = (
```
Comment (rianfowler) [inline web/src/features/AppConfig/components/AppConfig.tsx:487]: this is the bulk of the logic. we take the `ConfigGroup[]` and merge in the `ConfigGroupItemValidationErrors[]` by matching the field names from each array.

**Review by miaawong — APPROVED** [review]
great work! 🙌
